framework/src/engine/legacy/legacy_chain_handler.ts

```diff
@@ -12,56 +12,193 @@
+import { legacyChainBracketInfoSchema } from './schemas';
 
 interface LegacyChainHandlerArgs {
 	legacyConfig: LegacyConfig;
+	network: Network;
 }
 
 interface LegacyHandlerInitArgs {
-	db: Database | InMemoryDatabase;
+	db: Database;
 }
 
 export class LegacyChainHandler {
+	private readonly _network: Network;
+	private _storage!: Storage;
 	private readonly _legacyConfig: LegacyConfig;
-	private _db!: Database | InMemoryDatabase;
+	private _timeout!: NodeJS.Timeout;
 
 	public constructor(args: LegacyChainHandlerArgs) {
 		this._legacyConfig = args.legacyConfig;
+		this._network = args.network;
 	}
 
 	public async init(args: LegacyHandlerInitArgs): Promise<void> {
-		this._db = args.db;
-		const storage = new Storage(this._db as Database);
+		this._storage = new Storage(args.db);
 
 		for (const bracket of this._legacyConfig.brackets) {
-			const bracketInfo = await storage.getLegacyChainBracketInfo(
-				Buffer.from(bracket.snapshotBlockID, 'hex'),
-			);
-
-			if (!bracketInfo) {
-				await storage.setLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID), {
+			try {
+				await this._storage.getLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID, 'hex'));
+			} catch (err) {
+				// Save config brackets in advance, these will be used in next step (`sync`)
+				await this._storage.setLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID), {
 					startHeight: bracket.startHeight,
 					snapshotBlockHeight: bracket.snapshotHeight,
-					lastBlockHeight: bracket.startHeight,
+					lastBlockHeight: bracket.snapshotHeight,
 				});
 			}
 		}
 	}
 
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async syncBlocks(): Promise<void> {
-		// eslint-disable-next-line no-console
-		console.log(this._legacyConfig);
-		// eslint-disable-next-line no-console
-		console.log(this._db);
-		// TODO: implement the logic, general flow is as below
-		// 1. check config if sync legacy flag is true
-		// 2. Checks to what height blocks are present for a snapshotBlockID backwards
-		// 3. checks for all the peers node info if any peer has legacy blocks for the snapshot block ID, if not wait for X second
-		// 4. sync the blocks
-		// 5. Update node info
+	public async sync() {
+		if (!this._legacyConfig.sync) {
+			return;
+		}
+
+		for (const bracket of this._legacyConfig.brackets) {
+			const encodedBracketInfo = await this._storage.getLegacyChainBracketInfo(
+				Buffer.from(bracket.snapshotBlockID, 'hex'),
+			);
+			const bracketInfo = codec.decode<LegacyChainBracketInfo>(
+				legacyChainBracketInfoSchema,
+				encodedBracketInfo,
+			);
+			if (bracket.startHeight === bracketInfo.lastBlockHeight) {
+				continue;
+			}
+			const lastBlock = decodeBlock(
+				await this._storage.getBlockByHeight(bracketInfo.lastBlockHeight),
+			).block;
+
+			await this._trySyncBlocks(bracket, lastBlock);
+			await this._storage.setLegacyChainBracketInfo(
+				Buffer.from(bracket.snapshotBlockID, 'hex'),
+				bracketInfo,
```

Comment:
you need to update the bracket info with the last block after syncing which will be the block at the startHeight. Its better to set bracket everytime we finish one call with the peer, so once we process 100 blocks received, we should set bracket and then call another peer for next 100 blocks.